File: src/ssh_manager/core/expiry.py
"""Expiry policy engine - pure, no I/O.

A raw SSH keypair does not self-expire; this is a *policy reminder* we compute and
surface. From the inventory we derive, per key, ``expires_on`` and
``days_remaining`` and a ``state`` ∈ {ok, due_soon, overdue}. ``due_soon`` fires
once a key is within the warn window (the largest ``warn_before_days`` threshold).
"""
from __future__ import annotations

from dataclasses import dataclass
from datetime import date

from .inventory import Inventory, compute_expiry, is_archived_path

OK = "ok"
DUE_SOON = "due_soon"
OVERDUE = "overdue"
UNKNOWN = "unknown"


@dataclass(frozen=True)
class ExpiryStatus:
    fingerprint: str
    key_name: str
    profile: str
    created: str | None
    expires_on: str | None
    days_remaining: int | None
    state: str

    @property
    def needs_attention(self) -> bool:
        return self.state in (DUE_SOON, OVERDUE)


def _key_name(path: str) -> str:
    return path.rsplit("/", 1)[-1]


def compute_states(inventory: Inventory, *, warn_before_days: list[int],
                   today: date) -> list[ExpiryStatus]:
    """Per-key expiry status, sorted most-urgent first."""
    warn_window = max(warn_before_days) if warn_before_days else 30
    out: list[ExpiryStatus] = []
    for fp, rec in inventory.keys.items():
        if is_archived_path(rec.path):   # archived predecessor - not active, skip
            continue
        exp = rec.expires_on
        try:
            if exp is None and rec.created:
                exp = compute_expiry(rec.created, rec.rotate_after_days)
            parsed = date.fromisoformat(exp) if exp is not None else None
        except ValueError:
            # A hand-edited inventory with a malformed date shouldn't crash expiry/
            # audit - surface it as UNKNOWN instead of raising.
            parsed = None
        if parsed is None:
            out.append(ExpiryStatus(fp, _key_name(rec.path), rec.profile,
                                    rec.created, None, None, UNKNOWN))
            continue
        days = (parsed - today).days
        if days < 0:
            state = OVERDUE
        elif days <= warn_window:
            state = DUE_SOON
        else:
            state = OK
        out.append(ExpiryStatus(fp, _key_name(rec.path), rec.profile,
                                rec.created, exp, days, state))
    out.sort(key=lambda s: (s.days_remaining if s.days_remaining is not None else 10**9))
    return out


def cadence(states: list[ExpiryStatus]) -> str:
    """Notifier cadence: 'daily' once any key is in the warn window, else 'weekly'."""
    return "daily" if any(s.needs_attention for s in states) else "weekly"


def banner_lines(states: list[ExpiryStatus]) -> list[str]:
    """One ⚠ line per due/overdue key - the inline reminder."""
    lines: list[str] = []
    for s in states:
        if not s.needs_attention:
            continue
        if s.state == OVERDUE:
            when = f"OVERDUE by {abs(s.days_remaining or 0)} days"
        else:
            when = f"expires in {s.days_remaining} days"
        lines.append(
            f"⚠ {s.key_name} {when} ({s.expires_on}) - run: sshmgr rotate {s.key_name}"
        )
    return lines

File: internal/core/expiry/expiry.go
// Package expiry is the pure expiry-policy engine, ported from
// src/ssh_manager/core/expiry.py. A keypair does not self-expire; this computes
// the policy reminder (per key: expires_on, days_remaining, state).
package expiry

import (
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/simtabi/ssh-manager/internal/core/inventory"
)

// State values.
const (
	OK      = "ok"
	DueSoon = "due_soon"
	Overdue = "overdue"
	Unknown = "unknown"
)

const dateLayout = "2006-01-02"

// Status is one key's expiry policy state.
type Status struct {
	Fingerprint   string
	KeyName       string
	Profile       string
	Created       *string
	ExpiresOn     *string
	DaysRemaining *int
	State         string
}

// NeedsAttention is true for due_soon/overdue keys.
func (s Status) NeedsAttention() bool { return s.State == DueSoon || s.State == Overdue }

func keyName(path string) string {
	if i := strings.LastIndex(path, "/"); i >= 0 {
		return path[i+1:]
	}
	return path
}

// ComputeStates returns per-key expiry status, most-urgent first. today should be
// a date (UTC midnight); warnBeforeDays is the manifest's warn thresholds.
func ComputeStates(inv *inventory.Inventory, warnBeforeDays []int, today time.Time) []Status {
	warnWindow := 30
	if len(warnBeforeDays) > 0 {
		warnWindow = warnBeforeDays[0]
		for _, d := range warnBeforeDays[1:] {
			if d > warnWindow {
				warnWindow = d
			}
		}
	}
	var out []Status
	for fp, rec := range inv.Keys {
		if inventory.IsArchivedPath(rec.Path) {
			continue // archived predecessor, not active
		}
		exp := rec.ExpiresOn
		if exp == nil && rec.Created != nil && *rec.Created != "" {
			if e, err := inventory.ComputeExpiry(*rec.Created, rec.RotateAfterDays); err == nil {
				exp = &e
			}
		}
		var parsed *time.Time
		if exp != nil {
			if t, err := time.Parse(dateLayout, *exp); err == nil {
				parsed = &t
			} else {
				exp = nil // malformed -> unknown (don't crash)
			}
		}
		if parsed == nil {
			out = append(out, Status{
				Fingerprint: fp, KeyName: keyName(rec.Path), Profile: rec.Profile,
				Created: rec.Created, State: Unknown,
			})
			continue
		}
		days := int(parsed.Sub(today).Hours() / 24)
		state := OK
		switch {
		case days < 0:
			state = Overdue
		case days <= warnWindow:
			state = DueSoon
		}
		d := days
		out = append(out, Status{
			Fingerprint: fp, KeyName: keyName(rec.Path), Profile: rec.Profile,
			Created: rec.Created, ExpiresOn: exp, DaysRemaining: &d, State: state,
		})
	}
	sort.SliceStable(out, func(i, j int) bool {
		di, dj := daysOrInf(out[i].DaysRemaining), daysOrInf(out[j].DaysRemaining)
		if di != dj {
			return di < dj
		}
		return out[i].Fingerprint < out[j].Fingerprint // deterministic tiebreak
	})
	return out
}

func daysOrInf(d *int) int {
	if d == nil {
		return 1 << 30
	}
	return *d
}

// Cadence is "daily" once any key needs attention, else "weekly".
func Cadence(states []Status) string {
	for _, s := range states {
		if s.NeedsAttention() {
			return "daily"
		}
	}
	return "weekly"
}

// BannerLines is one warning line per due/overdue key - the inline reminder.
func BannerLines(states []Status) []string {
	var lines []string
	for _, s := range states {
		if !s.NeedsAttention() {
			continue
		}
		days := 0
		if s.DaysRemaining != nil {
			days = *s.DaysRemaining
		}
		var when string
		if s.State == Overdue {
			d := days
			if d < 0 {
				d = -d
			}
			when = fmt.Sprintf("OVERDUE by %d days", d)
		} else {
			when = fmt.Sprintf("expires in %d days", days)
		}
		exp := ""
		if s.ExpiresOn != nil {
			exp = *s.ExpiresOn
		}
		lines = append(lines, fmt.Sprintf(
			"⚠ %s %s (%s) - run: sshmgr rotate %s", s.KeyName, when, exp, s.KeyName))
	}
	return lines
}
